YouTrack API

kolbacer opened the pull request:
гет запросы в youtrack

kolbacer committed:
функции для получения инфы об ишьюсах

kolbacer committed:
тестовые запросы в YouTrack Глеба

kolbacer committed:
результаты теста

kolbacer commented:
.

Bezzzub commented on youtrack_api_get/test.py:
```diff
@@ -0,0 +1,14 @@
+from issue_info import *
+
+Gleb_token = "perm:cm9vdA==.NDYtMA==.yc2rmm8e2iecHUkPMBbKfXasd1fpx6"
+Gleb_youtrack_domain = "testbotproject"
+Gleb_issue_name = "DEMO-20"
+
+status = get_state(Gleb_youtrack_domain, Gleb_issue_name, Gleb_token)
+print(status)
+priority = get_priority(Gleb_youtrack_domain, Gleb_issue_name, Gleb_token)
+print(priority)
+timeEnd = get_time_end(Gleb_youtrack_domain, Gleb_issue_name, Gleb_token)
+print(timeEnd)
+timeStart = get_time_start(Gleb_youtrack_domain, Gleb_issue_name, Gleb_token)
+print(timeStart)
```
It's not a tests
Rename it to `example.py` and, if you want, write correct test via `pytest`

kolbacer replied:
done

Bezzzub commented on youtrack_api_get/issue_info.py:
```diff
@@ -0,0 +1,49 @@
+time_end_id = "117-0"
+
+
+def get_custom_field(youtrack_domain, issue_name, field_id, token):
```
Return `FieldData` entity, not just string

kolbacer replied:
(for me)
class FieldInfo(BaseModel):
    name: str
    value: str

kolbacer replied:
done

Bezzzub commented on youtrack_api_get/issue_info.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def get_custom_field(youtrack_domain, issue_name, field_id, token):
+    url = "https://" \
+          + youtrack_domain \
+          + ".myjetbrains.com/youtrack/api/issues/" \
+          + issue_name \
+          + "/customFields/" \
+          + field_id
```
Common url part move to class attribute.

kolbacer replied:
can't move to class attribute, but can into object attribute
(done)

Bezzzub commented on youtrack_api_get/issue_info.py:
```diff
@@ -0,0 +1,49 @@
+import httpx
+import datetime
+
+state_id = "111-2"
+priority_id = "111-0"
+time_end_id = "117-0"
```
Move to class attribute

kolbacer replied:
done

Bezzzub commented on youtrack_api_get/issue_info.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def get_custom_field(youtrack_domain, issue_name, field_id, token):
+    url = "https://" \
+          + youtrack_domain \
+          + ".myjetbrains.com/youtrack/api/issues/" \
+          + issue_name \
+          + "/customFields/" \
+          + field_id
```
Use `f-strings`   
f"https://{youtrack_domain}.myjetbrains.com/youtrack/api/issues/{issue_name}\{field_id}"

kolbacer replied:
done

Bezzzub commented on youtrack_api_get/issue_info.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def get_time_start(youtrack_domain, issue_name, token):
+    url = "https://" \
+               + youtrack_domain \
+               + ".myjetbrains.com/youtrack/api/issues/"\
+               + issue_name
```
Use `f-strings`

kolbacer replied:
done

Bezzzub commented on youtrack_api_get/issue_info.py:
```diff
@@ -0,0 +1,49 @@
+    headers = {'Authorization': 'Bearer ' + token}
+    params = {'fields': 'projectCustomField(field(name)),value(name)'}
+    response = httpx.get(url, headers=headers, params=params)
+    data = response.json()
```
Check the response. It is always valid?
```python
try:
    response.raise_for_status()
except HTTPError:
    do something
```

kolbacer replied:
done

Bezzzub commented on youtrack_api_get/issue_info.py:
```diff
@@ -0,0 +1,49 @@
+    name = "Created: "
+    time = data["created"]
+    value = datetime.datetime.fromtimestamp(time//1000.0)
+    return name + ": " + str(value)
```
Return entity, not just string

kolbacer replied:
done

kolbacer committed:
fixed get_time_end

оценочное время хранится в value(id), а не в value(name)
результат в формате 1н 1д 1ч 1м

kolbacer committed:
fixed get_time_end

нет, оно хранится в value(presentation)....

kolbacer committed:
Delete result.png

kolbacer committed:
Rename test.py to example_info.py

kolbacer committed:
Create issue_update.py

POST методы: 
update_priority - изменение приоритета
update_time_end - изменение планируемого времени выполнения задачи

kolbacer committed:
Create example_update.py

тестовые запросы в YouTrack Глеба (работает)

kolbacer committed:
rename folder

kolbacer committed:
Rename youtrack_api_get/example_update.py to youtrack_api/example_update.py

kolbacer committed:
Rename youtrack_api_get/issue_info.py to youtrack_api/issue_info.py

kolbacer committed:
Rename youtrack_api_get/issue_update.py to youtrack_api/issue_update.py

kolbacer committed:
Delete example_info.py

kolbacer committed:
Delete example_update.py

kolbacer committed:
Delete issue_info.py

kolbacer committed:
Delete issue_update.py

kolbacer committed:
Create IssueManager.py

- Mergerd issue_info and issue_update
- all methods in class "IssueManager"
- nested class "FieldInfo" implements results of requests
- request are checked
- f-strings are cool

kolbacer committed:
Test requests in Gleb's YouTrack

kolbacer committed:
results of test requests

kolbacer commented:
**all done, let's merge**

kolbacer committed:
Update IssueManager.py

FieldInfo moved to another file

kolbacer committed:
Create FieldInfo.py

this class implements results of requests

Bezzzub commented on youtrack_api/IssueManager.py:
```diff
@@ -0,0 +1,122 @@
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
+            return FieldInfo(None, None, False)
```
Возвращай None

kolbacer replied:
done

Bezzzub commented on youtrack_api/IssueManager.py:
```diff
@@ -0,0 +1,122 @@
+        try:
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
```
Return None

kolbacer replied:
done

Bezzzub commented on youtrack_api/IssueManager.py:
```diff
@@ -0,0 +1,122 @@
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
+            return FieldInfo(None, None, False)
```
None

kolbacer replied:
done

Kondrahin commented on youtrack_api/IssueManager.py:
```diff
@@ -0,0 +1,122 @@
+        try:
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
```
None

Bezzzub commented on youtrack_api/IssueManager.py:
```diff
@@ -0,0 +1,122 @@
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
+            return FieldInfo(None, None, False)
```
Return None

kolbacer replied:
DONE

Bezzzub commented on youtrack_api/IssueManager.py:
```diff
@@ -0,0 +1,122 @@
+            response.raise_for_status()
+        except httpx.HTTPError as e:
+            print(e)
+            return FieldInfo(None, None, False)
```
Return None

kolbacer replied:
DOOONNNEEEE

Bezzzub commented on youtrack_api/schemas/FieldInfo.py:
```diff
@@ -0,0 +1,11 @@
+class FieldInfo:
+    def __init__(self, name, value, is_valid=True):
```
Is_valid не нужно.

kolbacer replied:
ok

Bezzzub commented on youtrack_api/schemas/FieldInfo.py:
```diff
@@ -0,0 +1,11 @@
+        self.value = value  # not always string
+        self.is_valid = is_valid
+
+    def show(self):
```
Это тоже не нужно

kolbacer replied:
ок

Bezzzub commented on youtrack_api/schemas/FieldInfo.py:
```diff
@@ -0,0 +1,11 @@
+class FieldInfo:
```
Лучше отнаследоваться от `BaseModel` из [Pydantic](https://pydantic-docs.helpmanual.io/)
```python
from pydantic import BaseModel

class FieldInfo(BaseModel)
…
```

kolbacer replied:
сделал

kolbacer committed:
Update IssueManager.py

Noooooooooooooonnnneeeeeeeeeeeeeeeeeeee

kolbacer committed:
Update FieldInfo.py

inheritaned from BaseModel
smth with class attributes idk

kolbacer committed:
fixed example

kolbacer committed:
return none

kolbacer committed:
small fix

kolbacer committed:
I want to die
